### PR TITLE
Migrate UI codebase from C++/CX to C# | Part 1

### PR DESCRIPTION
**Pull request by tian-lt**

### CODE REVIEW

File: src/CalcViewModel/ApplicationViewModel.cpp [hunk 1]
```diff
@@ -21,7 +21,6 @@ using namespace Platform::Collections;
 using namespace std;
 using namespace Windows::System;
 using namespace Windows::Storage;
-using namespace Utils;
```
Comment (tian-lt) [inline src/CalcViewModel/ApplicationViewModel.cpp]: revert this change

File: src/CalcViewModel/Common/CopyPasteManager.cpp [hunk 2]
```diff
@@ -599,22 +599,22 @@ ULONG32 CopyPasteManager::ProgrammerOperandLength(Platform::String ^ operand, Nu
 }
 
 // return wstring after removing characters like space, comma, double quotes, and monetary prefix currency symbols supported by the Windows keyboard:
-// yen or yuan(¥) - 165
-// unspecified currency sign(¤) - 164
-// Ghanaian cedi(₵) - 8373
+// yen or yuan(Â¥) - 165
```
Comment (tian-lt) [inline src/CalcViewModel/Common/CopyPasteManager.cpp]: recover back these characters.

File: src/CalculatorUI/App.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+        {
+            public static readonly string AppMinWindowHeight = "AppMinWindowHeight";
+            public static readonly string AppMinWindowWidth = "AppMinWindowWidth";
+        }
+    }
+
+    /// <summary>
+    /// Provides application-specific behavior to supplement the default Application class.
+    /// </summary>
+    sealed partial class App : Application
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: Check C++/CX's generated code

File: src/CalculatorUI/App.xaml.cs [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+    {
+
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: Remove empty line

File: src/CalculatorUI/App.xaml.cs [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Controls;
+using Windows.UI.Xaml.Controls.Primitives;
+using Windows.UI.Xaml.Data;
+using Windows.UI.Xaml.Input;
+using Windows.UI.Xaml.Interop;
+using Windows.UI.Xaml.Media;
+using Windows.UI.Xaml.Media.Animation;
+using Windows.UI.Xaml.Navigation;
+//using Windows.ApplicationModel.Activation;
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: ?

File: src/CalculatorUI/App.xaml.cs [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+
+        /// <summary>
+        /// Initializes the singleton application object. This is the first line of authored code
+        /// executed, and as such is the logical equivalent of main() or WinMain().
+        /// </summary>
+        public App()
+        {
+            this.InitializeComponent();
+
+            m_preLaunched = false;
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs:67]: use the same format throughout entire project

File: src/CalculatorUI/App.xaml.cs [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+
+            // TODO: MSFT 14645325: Set this directly from XAML.
+            // Currently this is bugged so the property is only respected from code-behind.
+            this.HighContrastAdjustment = ApplicationHighContrastAdjustment.None;
+
+            this.Suspending += OnSuspending;
+
+            // CSHARP_MIGRATION: TODO:
+#if _DEBUG
+    this.DebugSettings.IsBindingTracingEnabled = true;
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: indents

File: src/CalculatorUI/App.xaml.cs [hunk 8]
```diff
@@ -0,0 +1,516 @@
...
+    this.DebugSettings.BindingFailed += new BindingFailedEventHandler((_In_ Object  /*sender*/, _In_ BindingFailedEventArgs  e) => {
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: take care of this block

File: src/CalculatorUI/App.xaml.cs [hunk 9]
```diff
@@ -0,0 +1,516 @@
...
+        if (IsDebuggerPresent())
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: Check compiling env

File: src/CalculatorUI/App.xaml.cs [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+        {
+            string  errorMessage = e.Message;
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: handle extra spaces

File: src/CalculatorUI/App.xaml.cs [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+#endif
+        }
+
+        /// <summary>
+        /// Invoked when the application is launched normally by the end user. Other entry points
+        /// will be used when the application is launched to open a specific file, to display
+        /// search results, and so forth.
+        /// </summary>
+        /// <param name="e">Details about the launch request and process.</param>
+        protected override void OnLaunched(LaunchActivatedEventArgs args)
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs:96]: Double check C# virtual and override keywords

File: src/CalculatorUI/App.xaml.cs [hunk 12]
```diff
@@ -0,0 +1,516 @@
...
+            {
+            }
+        }
+
+        private async Task HandleViewReleaseAndRemoveWindowFromMap(WindowFrameService frameService)
+        {
+            WeakReference weak = new WeakReference(this);
+
+            // Unregister the event handler of the Main Page
+            var frame = (Window.Current.Content as Frame);
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs:463]: Investigate safe_cast

File: src/CalculatorUI/App.xaml.cs [hunk 13]
```diff
@@ -0,0 +1,516 @@
...
+            var mainPage = (frame.Content as MainPage);
+
+            // CSHARP_MIGRATION: TODO:
+            //mainPage.UnregisterEventHandlers();
+
+            await frameService.HandleViewRelease();
+            await Task.Run(() =>
+            {
+                var that = weak.Target as App;
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs:472]: Check why weak ref has been used here

File: src/CalculatorUI/App.xaml.cs [hunk 14]
```diff
@@ -0,0 +1,516 @@
...
+                else
+                {
+                    return null;
+                }
+            }
+        }
+
+        private void RemoveWindowFromMap(int viewId)
+        {
+
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: Empty line

File: src/CalculatorUI/App.xaml.cs [hunk 15]
```diff
@@ -0,0 +1,516 @@
...
+            lock (m_windowsMapLockMutex)
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: Check read write locks in C#

File: src/CalculatorUI/App.xaml.cs [hunk 16]
```diff
@@ -0,0 +1,516 @@
...
+            {
+                WindowFrameService iter;
+                if (m_secondaryWindows.TryGetValue(viewId, out iter))
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: Get value directly

File: src/CalculatorUI/App.xaml.cs [hunk 17]
```diff
@@ -0,0 +1,516 @@
...
+                {
+                    m_secondaryWindows.Remove(viewId);
+                }
+                else
+                {
+                    Debug.Assert(iter != null, "Window does not exist in the list");
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: Rewrite this condition

File: src/CalculatorUI/App.xaml.cs [hunk 18]
```diff
@@ -0,0 +1,516 @@
...
+            appView.SetPreferredMinSize(minWindowSize);
+
+            // Place the frame in the current Window
+            Window.Current.Content = rootFrame;
+            Window.Current.Activate();
+        }
+
+        private void OnAppLaunch(IActivatedEventArgs args, String argument)
+        {
+            // CSHARP_MIGRATION: TODO:
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: Remove this

File: src/CalculatorUI/App.xaml.cs [hunk 19]
```diff
@@ -0,0 +1,516 @@
...
+            // Uncomment the following lines to display frame-rate and per-frame CPU usage info.
+            //#if _DEBUG
+            //    if (IsDebuggerPresent())
+            //    {
+            //        DebugSettings->EnableFrameRateCounter = true;
+            //    }
+            //#endif
+
+            args.SplashScreen.Dismissed += DismissedEventHandler;
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs:165]: Check typedeventhandler types

File: src/CalculatorUI/App.xaml.cs [hunk 20]
```diff
@@ -0,0 +1,516 @@
...
+            WeakReference weak = new WeakReference(this);
+
+            float minWindowWidth = (float)((double)Resources[ApplicationResourceKeys.Globals.AppMinWindowWidth]);
+            float minWindowHeight = (float)((double)Resources[ApplicationResourceKeys.Globals.AppMinWindowHeight]);
+            Size minWindowSize = SizeHelper.FromDimensions(minWindowWidth, minWindowHeight);
+
+            ApplicationView appView = ApplicationView.GetForCurrentView();
+            ApplicationDataContainer localSettings = ApplicationData.Current.LocalSettings;
+            // For very first launch, set the size of the calc as size of the default standard mode
+            if (!localSettings.Values.ContainsKey("VeryFirstLaunch"))
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs:177]: Check the original type of `Values`

File: src/CalculatorUI/App.xaml.cs [hunk 21]
```diff
@@ -0,0 +1,516 @@
...
+                rootFrame = App.CreateFrame();
+
+                // When the navigation stack isn't restored navigate to the first page,
+                // configuring the new page by passing required information as a navigation
+                // parameter
+                if (!rootFrame.Navigate(typeof(MainPage), argument))
+                {
+                    // We couldn't navigate to the main page, kill the app so we have a good
+                    // stack to debug
+                    throw new Exception();
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: what was std::bad_exception being used for?

Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs]: Maybe we need to improve this exception.

Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs]: SystemException ?

File: src/CalculatorUI/App.xaml.cs [hunk 22]
```diff
@@ -0,0 +1,516 @@
...
+                m_mainViewId = ApplicationView.GetForCurrentView().Id;
+                AddWindowToMap(WindowFrameService.CreateNewWindowFrameService(rootFrame, false, weak));
+            }
+            else
+            {
+                // For first launch, LaunchStart is logged in constructor, this is for subsequent launches.
+
+                // !Phone check is required because even in continuum mode user interaction mode is Mouse not Touch
+                if ((UIViewSettings.GetForCurrentView().UserInteractionMode == UserInteractionMode.Mouse)
+                    & (!Windows.Foundation.Metadata.ApiInformation.IsTypePresent("Windows.Phone.UI.Input.HardwareButtons")))
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: Check what happened when casting && to &

File: src/CalculatorUI/App.xaml.cs [hunk 23]
```diff
@@ -0,0 +1,516 @@
...
+                {
+                    // If the pre-launch hasn't happened then allow for the new window/view creation
+                    if (!m_preLaunched)
+                    {
+                        var newCoreAppView = CoreApplication.CreateNewView();
+                        _ = newCoreAppView.Dispatcher.RunAsync(
+                            CoreDispatcherPriority.Normal, new DispatchedHandler(async () =>
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: if DispatcheHandler is needed?

File: src/CalculatorUI/App.xaml.cs [hunk 24]
```diff
@@ -0,0 +1,516 @@
...
+                            {
+                                if (weak.IsAlive)
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: Remove this

File: src/CalculatorUI/App.xaml.cs [hunk 25]
```diff
@@ -0,0 +1,516 @@
...
+                                {
+                                    var that = weak.Target as App;
+                                    var newRootFrame = App.CreateFrame();
+
+                                    SetMinWindowSizeAndActivate(newRootFrame, minWindowSize);
+
+                                    if (!rootFrame.Navigate(typeof(MainPage), argument))
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: Rename rootFrame to newRootFrame

Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs]: try shadow if we can do so

File: src/CalculatorUI/App.xaml.cs [hunk 26]
```diff
@@ -0,0 +1,516 @@
...
+                                    {
+                                        // We couldn't navigate to the main page, kill the app so we have a good
+                                        // stack to debug
+                                        throw new Exception();
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: Same

File: src/CalculatorUI/App.xaml.cs [hunk 27]
```diff
@@ -0,0 +1,516 @@
...
+                                    }
+
+                                    var frameService = WindowFrameService.CreateNewWindowFrameService(rootFrame, true, weak);
+                                    that.AddWindowToMap(frameService);
+
+                                    var dispatcher = CoreWindow.GetForCurrentThread().Dispatcher;
+                                    var safeFrameServiceCreation = new SafeFrameWindowCreation(frameService, that);
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: RAII?

File: src/CalculatorUI/App.xaml.cs [hunk 28]
```diff
@@ -0,0 +1,516 @@
...
+                                    ActivationViewSwitcher activationViewSwitcher = null;
+                                    var activateEventArgs = (args as IViewSwitcherProvider);
+                                    if (activateEventArgs != null)
+                                    {
+                                        activationViewSwitcher = activateEventArgs.ViewSwitcher;
+                                    }
+
+                                    if (activationViewSwitcher != null)
+                                    {
+                                        await activationViewSwitcher.ShowAsStandaloneAsync(newWindowId, ViewSizePreference.Default);
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: _ =

File: src/CalculatorUI/App.xaml.cs [hunk 29]
```diff
@@ -0,0 +1,516 @@
...
+                                        safeFrameServiceCreation.SetOperationSuccess(true);
+                                    }
+                                    else
+                                    {
+                                        var activatedEventArgs = (args as IApplicationViewActivatedEventArgs);
+                                        if ((activatedEventArgs != null) && (activatedEventArgs.CurrentlyShownApplicationViewId != 0))
+                                        {
+                                            var viewShown = await ApplicationViewSwitcher.TryShowAsStandaloneAsync(
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: Task.Run() and ContinueWith

File: src/CalculatorUI/App.xaml.cs [hunk 30]
```diff
@@ -0,0 +1,516 @@
...
+            //NarratorNotifier.RegisterDependencyProperties();
+        }
+
+        private void OnSuspending(Object sender, SuspendingEventArgs args)
+        {
+            // CSHARP_MIGRATION: TODO:
+            //TraceLogger.GetInstance().LogButtonUsage();
+        }
+
+        sealed class SafeFrameWindowCreation
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: Re-implementation need?

File: src/CalculatorUI/App.xaml.cs (hunk 29 above)
Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs]: SafeFrameServiceCreate problem requires here to using await statements to translate the code.

File: src/CalculatorUI/App.xaml.cs (hunk 5 above)
Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs]: `using Windows.ApplicationModel.Activation` has appeared at above within the file.
It should be removed.

File: src/CalculatorUI/App.xaml.cs (hunk 9 above)
Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs]: `IsDebuggerPresent()` is an API provided by debugapi.h from Windows 10 SDK.

```
WINBASEAPI
BOOL
WINAPI
IsDebuggerPresent(
    VOID
    );
```

It should be reinterpreted to `System.Diagnostics.Debugger.IsAttached` regarding [the same quastion being asked in stackoverflow](https://stackoverflow.com/questions/12422242/is-there-a-c-sharp-equivalent-to-isdebuggerpresent)

File: src/CalculatorUI/App.xaml.cs (hunk 22 above)
Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs]: There's a bug in Sharpifier.
```
private static string MigratePointers(string content)
{
...
    content = content.ReplaceRegex(@"(?<=(var|return|=|\n|>|\[|\()[ \t]*)[\*&]", "");
}
```
The regex rule will match `&&` when it's at the beginning of line.

File: src/CalculatorUI/App.xaml.cs (hunk 6 above)
Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs:67]: will leverage automations to ensure the project has same format of the code.

File: src/CalculatorUI/App.xaml.cs (hunk 10 above)
Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs]: same as above, will fix this by leveraging automations later.

File: src/CalculatorUI/App.xaml.cs (hunk 11 above)
Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs:96]: Regarding [this article](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/virtual)

> You cannot use the `virtual` modifier with the `static`, `abstract`, `private`, or `override` modifiers.

File: src/CalculatorUI/App.xaml.cs (hunk 16 above)
Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs]: Fixed, check it out in the latest commit.

File: src/CalculatorUI/App.xaml.cs (hunk 17 above)
Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs]: Fixed, check it out in the latest commit.

File: src/CalculatorUI/App.xaml.cs (hunk 30 above)
Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs]: `SafeFrameWindowCreation` has been refactored into a `IDisposable` class.
See more comments inline

File: src/CalculatorUI/App.xaml.cs (hunk 23 above)
Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs]: Removed DispatcheHandler to make the C# code more clean

File: src/CalculatorUI/App.xaml.cs (hunk 3 above)
Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs]: in App.g.h
```
    partial ref class App :  public ::Windows::UI::Xaml::Application,
        public ::Windows::UI::Xaml::Markup::IXamlMetadataProvider
```

Looks like C++/CX's `partial` keyword can be used partially? 🤔

Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs]: Regarding [this article](https://docs.microsoft.com/en-us/cpp/extensions/partial-cpp-component-extensions?view=msvc-160)

> For a ref class that has two partial definitions, the partial keyword is applied to the first occurrence of the definition, and this is typically done by auto-generated code, so that a human coder doesn't use the keyword very often

File: src/CalculatorUI/App.xaml.cs (hunk 20 above)
Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs:177]: Regarding [this article](https://docs.microsoft.com/en-us/uwp/api/windows.foundation.collections.ipropertyset?view=winrt-19041)

`ApplicationDataContainer` is a `IPropertySet` Interface which derived from `IDictionary`

```
[Windows.Foundation.Metadata.ContractVersion(typeof(Windows.Foundation.FoundationContract), 65536)]
[Windows.Foundation.Metadata.Guid(2319707551, 62694, 17441, 172, 249, 29, 171, 41, 134, 130, 12)]
[Windows.Foundation.Metadata.HasVariant]
public interface IPropertySet : IDictionary<string,object>, IEnumerable<KeyValuePair<string,object>>, IObservableMap<string,object>
```

>This code snippet comes from the ControlChannelTrigger HTTP client sample. Note how the C# version casts to IDictionary<string,object> so that it can use the indexer, whereas the Visual C++ component extensions (C++/CX) version uses HasKey and Lookup.

File: src/CalculatorUI/App.xaml.cs (hunk 21 above)
Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs]: Changed this into `throw new SystemException()`

File: src/CalculatorUI/Common/AppLifecycleLogger.cs [hunk 31]
```diff
@@ -0,0 +1,167 @@
...
+
+        public void ResumeVisibleComplete()
+        {
+
+            if (!GetTraceLoggingProviderEnabled())
+                return;
+
+            LoggingFields fields = new LoggingFields();
+            PopulateAppInfo(fields);
+            LogAppLifecycleEvent("ModernAppResume_VisibleComplete", fields);
```
Comment (wusixu) [inline src/CalculatorUI/Common/AppLifecycleLogger.cs:93]: LaunchVisibleComplete（), ResumeUIResponsive(), ResumeVisibleComplete(), etc have the same logic except the log event name, encapsulate this logic into a method.

File: src/CalculatorUI/Common/AppLifecycleLogger.cs [hunk 32]
```diff
@@ -0,0 +1,167 @@
...
+
+        public void ResizeVisibleComplete(int viewId)
+        {
+            if (!GetTraceLoggingProviderEnabled())
+                return;
+
+            LoggingFields fields = new LoggingFields();
+            PopulateAppInfo(fields);
+            fields.AddInt32("ViewId", viewId);
+            LogAppLifecycleEvent("ModernAppResize_VisibleComplete", fields);
```
Comment (wusixu) [inline src/CalculatorUI/Common/AppLifecycleLogger.cs:126]: Same with  ResizeUIResponsive(int viewId)

File: src/CalculatorUI/Common/AppLifecycleLogger.cs [hunk 33]
```diff
@@ -0,0 +1,167 @@
...
+        // TraceLoggingKeyword(MICROSOFT_KEYWORD_LEVEL_3) accordingly c) Should accept a variable number of additional data arguments if needed
+        private void LogAppLifecycleEvent(string eventName, LoggingFields fields)
+        {
+            m_appLifecycleProvider.LogEvent(
+                eventName, fields, LoggingLevel.Information, new LoggingOptions(Globals.MICROSOFT_KEYWORD_LEVEL_3 | WinMeta.WINEVENT_KEYWORD_RESPONSE_TIME));
+        }
+
+        private void PopulateAppInfo(LoggingFields fields)
+        {
+
```
Comment (wusixu) [inline src/CalculatorUI/Common/AppLifecycleLogger.cs:151]: Remove empty lines

File: src/CalculatorUI/Common/AppLifecycleLogger.cs [hunk 34]
```diff
@@ -0,0 +1,167 @@
...
+            var appId = CoreApplication.Id;
+            var aumId = Package.Current.Id.FamilyName + "!" + appId;
+            var packageFullName = Package.Current.Id.FullName;
+            var psmKey = Package.Current.Id.FullName + "+" + appId;
+
+            fields.AddString("AumId", aumId);
```
Comment (wusixu) [inline src/CalculatorUI/Common/AppLifecycleLogger.cs:157]: Is "AumId" a key? if it is, it's better to set these keys as constant instead of hard code.

File: src/CalculatorUI/Utils/DispatcherTimerDelayer.cs [hunk 35]
```diff
@@ -0,0 +1,47 @@
...
+{
+    public sealed class DispatcherTimerDelayer
+    {
+        public event EventHandler<object> Action;
+
+        public DispatcherTimerDelayer(TimeSpan timeSpan)
+        {
+            m_timer = new DispatcherTimer();
+            m_timer.Interval = timeSpan;
+            var interval = m_timer.Interval;
```
Comment (wusixu) [inline src/CalculatorUI/Utils/DispatcherTimerDelayer.cs:17]: What is this for?

File: src/CalculatorUI/Utils/VisualTree.cs [hunk 36]
```diff
@@ -0,0 +1,136 @@
...
+        }
+
+        /// <summary>
+        /// Find first descendant control of a specified type.
+        /// </summary>
+        /// <param name="element">Parent element.</param>
+        /// <param name="type">Type of descendant.</param>
+        /// <returns>Descendant control or null if not found.</returns>
+        private static DependencyObject FindDescendant(DependencyObject element, Type typeName)
+        {
```
Comment (wusixu) [inline src/CalculatorUI/Utils/VisualTree.cs:58]: Seem that the logic is pretty the same as FindDescendantByName, then should we also need the validation like line 27?

Comment (wusixu) [reply src/CalculatorUI/Utils/VisualTree.cs:58]: Nit: Since the logic of finding by type name is similar to by name, we can encapsulate it as a method and pass the core caparison as a lambda parameter.

File: src/CalculatorUI/WindowFrameService.cs [hunk 37]
```diff
@@ -0,0 +1,207 @@
...
+        }
+
+        public void InvokeWindowClosingHandlers()
+        {
+            // Should be called only once just before we kill the window.
+            foreach (var handler in m_onWindowClosingHandlers)
+            {
+                handler();
+            }
+            m_onWindowClosingHandlers.Clear();
```
Comment (wusixu) [inline src/CalculatorUI/WindowFrameService.cs:136]: Is this necessary?

File: src/CalculatorUI/App.xaml.cs [hunk 38]
```diff
@@ -0,0 +1,537 @@
...
+        {
+            public static readonly string AppMinWindowHeight = "AppMinWindowHeight";
+            public static readonly string AppMinWindowWidth = "AppMinWindowWidth";
+        }
+    }
+
+    /// <summary>
+    /// Provides application-specific behavior to supplement the default Application class.
+    /// </summary>
+    sealed partial class App : Application
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: Remove : Application

File: src/CalculatorUI/App.xaml.cs (hunk 12 above)
Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs:463]: Investigate static_cast

File: src/CalculatorUI/App.xaml.cs [hunk 39]
```diff
@@ -0,0 +1,537 @@
...
+                m_windowsMapLock.ExitReadLock();
+            }
+        }
+
+        private void RemoveWindowFromMap(int viewId)
+        {
+            m_windowsMapLock.EnterWriteLock();
+            try
+            {
+                Debug.Assert(m_secondaryWindows.ContainsKey(viewId), "Window does not exist in the list");
```
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: Use the result returned by Remove method

File: src/CalculatorUI/App.xaml.cs (hunk 30 above)
Comment (tian-lt) [inline src/CalculatorUI/App.xaml.cs]: add `private`

File: src/CalculatorUI/App.xaml.cs (hunk 39 above)
Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs]: resolved

File: src/CalculatorUI/Common/AppLifecycleLogger.cs (hunk 31 above)
Comment (tian-lt) [reply src/CalculatorUI/Common/AppLifecycleLogger.cs:93]: Thanks for the comments. These are great points to fix.
However, we'd like schedule such optimization works after migrating C++/CX to C# successfully since we don't want to introduce refactoring risks into the migration works.

File: src/CalculatorUI/Common/AppLifecycleLogger.cs (hunk 33 above)
Comment (tian-lt) [reply src/CalculatorUI/Common/AppLifecycleLogger.cs:151]: Same as above.
Will resolve formatting problems later together by leveraging a proper automation.

File: src/CalculatorUI/Common/AppLifecycleLogger.cs (hunk 34 above)
Comment (tian-lt) [reply src/CalculatorUI/Common/AppLifecycleLogger.cs:157]: same as above

File: src/CalculatorUI/Utils/DispatcherTimerDelayer.cs (hunk 35 above)
Comment (tian-lt) [reply src/CalculatorUI/Utils/DispatcherTimerDelayer.cs:17]: Won't fix. same as above

File: src/CalculatorUI/Utils/VisualTree.cs (hunk 36 above)
Comment (tian-lt) [reply src/CalculatorUI/Utils/VisualTree.cs:58]: Won't fix. Same as above

File: src/CalculatorUI/WindowFrameService.cs (hunk 37 above)
Comment (tian-lt) [reply src/CalculatorUI/WindowFrameService.cs:136]: Won't fix.
Same as above

File: src/CalculatorUI/App.xaml.cs (hunk 7 above)
Comment (tian-lt) [reply src/CalculatorUI/App.xaml.cs]: fixed

**Comment by tian-lt** [comment]
branch `csharpui-sharpify` will become very unstable since we decided to use to sync code that is in developing state.
So, here I close this PR without merging it into target branch, and later we will host new code review sessions for entire code changes.
